agent/discover/hanacluster.go

```diff
@@ -0,0 +1,53 @@
+package discover
+
+import (
+	"fmt"
+
+	consul "github.com/hashicorp/consul/api"
+	consul_internal "github.com/trento-project/trento/internal/consul"
+)
+
+type HanaClusterDiscover struct {
+	cluster               ClusterDiscover
+	SAPid                 string
+	SAPHanaInstanceNumber string
+}
+
+// check if the current node this trento agent is running on can be discovered
+// by HanaClusterDiscover, with other words is a SAP Hana Pacemaker cluster
+func (hana HanaClusterDiscover) shouldDiscover(client consul.Client) bool {
+
+	return len(hana.SAPid) > 0
+}
+
+// Provide a Consul Key-Value Store Prefix that is used to store individual Key-Values for this Cluster
+func (hana HanaClusterDiscover) KVStorePrefix() string {
+	return fmt.Sprintf(
+		"%s/%s",
+		consul_internal.KV_CLUSTERS_PATH, hana.cluster.clusterName)
+}
+
+// Discovery module handling specifics for SAP Hana clusters
+func NewHanaClusterDiscover(client consul.Client) Discovery {
```

Comment:
The name of this constructor is quite misleading: it would suggest that it would return a `HanaClusterDiscover` type, but it actually returns a `Discovery` function. 

By trying to understand how these two types are composed together, it appears to me that creating any `Discovery` type will need some `Discoverer` interface, so maybe we actually need a companion constructor to build the `Discovery` function by injecting a `Discoverer` in it?

All that said, I would go back to my initial suggestion to maybe collapse the `Discovery` function type into the `Discoverer` interface as a method, may this would simplify things a bit?
To be honest, I still don't quite get why we have two different types instead of one single `Discovery` interface, with something like a `Discovery.run()` method invoked by the Agent loop; am I missing something?

Reply:
I was following a guideline on the web on how to implement polymorphism using closures in Golang. I've now changed it as you requested. it is significantly more code duplication, but hopefully it'll get the PR merged then to be able to move forward.